compiler/rustc_resolve/src/def_collector.rs

```diff
@@ -265,16 +268,24 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
             ExprKind::Closure(_, _, asyncness, ..) => {
                 // Async closures desugar to closures inside of closures, so
                 // we must create two defs.
-                let closure_def = self.create_def(expr.id, DefPathData::ClosureExpr, expr.span);
+                let closure_def =
+                    self.create_def(expr.id, DefPathData::ClosureExpr(None), expr.span);
```

Comment:
The closure could still be a regular generator if the body contains a `yield`.
It should be included here, otherwise we'll have an inconsistency between the `DefPath` and HIR.